File: src/executor.js
/**
 * Executor for TradeMinutes DSL
 * Executes the parsed AST and runs commands
 */

class Executor {
  constructor() {
    this.variables = {};
    this.stepResults = {};
  }

  /**
   * Execute a parsed AST
   * @param {Object} ast - The parsed Abstract Syntax Tree
   */
  execute(ast) {
    console.log('🚀 Executing DSL Workflow...\n');
    
    // Execute variables first
    if (ast.variables && ast.variables.length > 0) {
      console.log('📝 Setting up variables...');
      ast.variables.forEach(variable => {
        this.executeVariable(variable);
      });
      console.log('');
    }
    
    // Execute workflows
    ast.workflows.forEach(workflow => {
      this.executeWorkflow(workflow);
    });
  }

  /**
   * Execute a variable declaration
   */
  executeVariable(variable) {
    const value = this.evaluateExpression(variable.value);
    this.variables[variable.name] = value;
    console.log(`  ✅ Set ${variable.declarationType} ${variable.name} = ${JSON.stringify(value)}`);
  }

  /**
   * Execute a workflow
   */
  executeWorkflow(workflow) {
    console.log(`📋 Executing workflow: ${workflow.name}`);
    console.log(`   Steps: ${workflow.steps.length}\n`);
    
    workflow.steps.forEach((step, index) => {
      console.log(`   Step ${index + 1}:`);
      this.executeStep(step);
      console.log('');
    });
  }

  /**
   * Execute a step
   */
  executeStep(step) {
    if (step.type === 'Step') {
      this.executeCommand(step);
    } else if (step.type === 'ConditionalStatement') {
      this.executeConditional(step);
    }
  }

  /**
   * Execute a command
   */
  executeCommand(step) {
    const command = step.command;
    const args = command.arguments.map(arg => this.evaluateExpression(arg));
    
    console.log(`     Command: ${command.name}`);
    console.log(`     Arguments: ${args.map(arg => JSON.stringify(arg)).join(', ')}`);
    
    // Execute the command
    switch (command.name) {
      case 'print':
      case 'log':
        console.log(`     📤 OUTPUT: ${args[0]}`);
        this.stepResults[step.number] = { type: 'print', output: args[0] };
        break;
        
      case 'fetch':
        console.log(`     🌐 FETCHING: ${args[0]}`);
        this.stepResults[step.number] = { 
          type: 'fetch', 
          url: args[0], 
          status: 'success',
          data: { message: 'Mock data from API' }
        };
        break;
        
      case 'send_email':
        console.log(`     📧 SENDING EMAIL: ${args[0]}`);
        this.stepResults[step.number] = { 
          type: 'send_email', 
          to: args[0], 
          data: args[1],
          status: 'sent'
        };
        break;
        
      case 'notify':
        console.log(`     🔔 NOTIFICATION: ${args[0]} - ${args[1]}`);
        this.stepResults[step.number] = { 
          type: 'notify', 
          to: args[0], 
          message: args[1],
          status: 'sent'
        };
        break;
        
      default:
        console.log(`     ⚙️  Executing: ${command.name}`);
        this.stepResults[step.number] = { 
          type: command.name, 
          status: 'completed',
          data: args
        };
    }
  }

  /**
   * Execute a conditional statement
   */
  executeConditional(conditional) {
    console.log(`     🔀 Conditional Statement:`);
    
    const conditionResult = this.evaluateCondition(conditional.condition);
    console.log(`     Condition: ${conditionResult ? '✅ TRUE' : '❌ FALSE'}`);
    
    if (conditionResult) {
      console.log(`     Executing IF branch (${conditional.ifSteps.length} steps):`);
      conditional.ifSteps.forEach((step, index) => {
        console.log(`       IF Step ${index + 1}:`);
        this.executeStep(step);
      });
    } else {
      console.log(`     Executing ELSE branch (${conditional.elseSteps.length} steps):`);
      conditional.elseSteps.forEach((step, index) => {
        console.log(`       ELSE Step ${index + 1}:`);
        this.executeStep(step);
      });
    }
  }

  /**
   * Evaluate a condition
   */
  evaluateCondition(condition) {
    if (condition.type === 'ComparisonExpression') {
      const left = this.evaluateExpression(condition.left);
      const right = this.evaluateExpression(condition.right);
      
      switch (condition.operator) {
        case '==': return left == right;
        case '!=': return left != right;
        case '>': return left > right;
        case '<': return left < right;
        case '>=': return left >= right;
        case '<=': return left <= right;
        default: return false;
      }
    }
    return false;
  }

  /**
   * Evaluate an expression
   */
  evaluateExpression(expression) {
    if (expression.type === 'StringLiteral') {
      return expression.value;
    } else if (expression.type === 'NumberLiteral') {
      return expression.value;
    } else if (expression.type === 'Identifier') {
      return this.variables[expression.value];
    } else if (expression.type === 'BinaryExpression') {
      const left = this.evaluateExpression(expression.left);
      const right = this.evaluateExpression(expression.right);
      
      if (expression.operator === '+') {
        return left + right;
      }
    } else if (expression.type === 'PropertyAccess') {
      const object = this.evaluateExpression(expression.object);
      if (object && object.status) {
        return object.status;
      }
    } else if (expression.type === 'StepReference') {
      return this.stepResults[expression.stepNumber];
    }
    
    return expression.value || expression;
  }
}

module.exports = Executor; 

File: rust-version/src/executor.rs
use anyhow::{anyhow, Result};
use std::collections::HashMap;
use crate::ast::*;

#[derive(Debug, Clone)]
pub struct StepResult {
    pub success: bool,
    pub data: String,
    pub status: u32,
    pub message: String,
}

impl StepResult {
    pub fn new(success: bool, data: String, status: u32, message: String) -> Self {
        StepResult {
            success,
            data,
            status,
            message,
        }
    }
}

pub struct Executor {
    variables: HashMap<String, String>,
    step_results: HashMap<u32, StepResult>,
}

impl Executor {
    pub fn new() -> Self {
        Executor {
            variables: HashMap::new(),
            step_results: HashMap::new(),
        }
    }
    
    pub fn execute(&mut self, program: &Program) -> Result<()> {
        println!("🚀 Executing TradeMinutes DSL Program");
        println!("=====================================");
        
        // Execute variable declarations
        for variable in &program.variables {
            self.execute_variable(variable)?;
        }
        
        // Execute workflows
        for workflow in &program.workflows {
            self.execute_workflow(workflow)?;
        }
        
        Ok(())
    }
    
    fn execute_variable(&mut self, variable: &VariableDeclaration) -> Result<()> {
        let value = self.evaluate_expression(&variable.value)?;
        self.variables.insert(variable.name.clone(), value);
        println!("📦 Variable '{}' = '{}'", variable.name, self.variables[&variable.name]);
        Ok(())
    }
    
    fn execute_workflow(&mut self, workflow: &Workflow) -> Result<()> {
        println!("\n🔄 Executing workflow: {}", workflow.name);
        
        // Execute workflow variables first
        for variable in &workflow.variables {
            self.execute_variable(variable)?;
        }
        
        for step in &workflow.steps {
            self.execute_step(step)?;
        }
        
        Ok(())
    }
    
    fn execute_step(&mut self, step: &Step) -> Result<()> {
        println!("  📋 Step {}: ", step.id);
        
        match &step.content {
            StepContent::Command(command) => {
                self.execute_command(step.id, command)?;
            }
            StepContent::Conditional(conditional) => {
                self.execute_conditional(conditional)?;
            }
        }
        
        Ok(())
    }
    
    fn execute_command(&mut self, step_id: u32, command: &Command) -> Result<()> {
        let args: Vec<String> = command.arguments
            .iter()
            .map(|expr| self.evaluate_expression(expr))
            .collect::<Result<Vec<String>>>()?;
        
        match command.name.as_str() {
            "print" => {
                let message = args.join(" ");
                println!("    📤 Print: {}", message);
                self.step_results.insert(step_id, StepResult::new(
                    true, message, 200, "Print executed successfully".to_string()
                ));
            }
            "log" => {
                let message = args.join(" ");
                println!("    📝 Log: {}", message);
                self.step_results.insert(step_id, StepResult::new(
                    true, message, 200, "Log executed successfully".to_string()
                ));
            }
            "fetch" => {
                let default_url = "https://api.example.com".to_string();
                let url = args.get(0).unwrap_or(&default_url);
                println!("    🌐 Fetch: {}", url);
                // Simulate fetch result
                let result = StepResult::new(
                    true,
                    format!("{{\"data\": \"Sample data from {}\"}}", url),
                    200,
                    "Fetch completed successfully".to_string()
                );
                self.step_results.insert(step_id, result);
            }
            "send_email" => {
                let default_to = "user@example.com".to_string();
                let default_subject = "Notification".to_string();
                let to = args.get(0).unwrap_or(&default_to);
                let subject = args.get(1).unwrap_or(&default_subject);
                println!("    📧 Send Email: {} - {}", to, subject);
                self.step_results.insert(step_id, StepResult::new(
                    true, format!("Email sent to {}", to), 200, "Email sent successfully".to_string()
                ));
            }
            "notify" => {
                let message = args.join(" ");
                println!("    🔔 Notify: {}", message);
                self.step_results.insert(step_id, StepResult::new(
                    true, message, 200, "Notification sent successfully".to_string()
                ));
            }
            // AI-specific commands for workflow integration
            "input" => {
                let variable_name = args.get(0).unwrap_or(&"user_input".to_string()).clone();
                let input_type = args.get(1).unwrap_or(&"text".to_string()).clone();
                let placeholder = args.get(2).unwrap_or(&"Enter value".to_string()).clone();
                println!("    📝 Input: Collect '{}' as {} ({})", variable_name, input_type, placeholder);
                
                // Simulate user input collection
                let result = StepResult::new(
                    true,
                    format!("{{\"variable\": \"{}\", \"type\": \"{}\", \"placeholder\": \"{}\"}}", 
                           variable_name, input_type, placeholder),
                    200,
                    "Input collected successfully".to_string()
                );
                self.step_results.insert(step_id, result);
            }
            "generate" => {
                let prompt = args.get(0).unwrap_or(&"Generate content".to_string()).clone();
                let model = args.get(1).unwrap_or(&"mistral-small-latest".to_string()).clone();
                let temperature = args.get(2).unwrap_or(&"0.7".to_string()).clone();
                println!("    🤖 Generate: Using {} (temp: {}) with prompt: '{}'", model, temperature, prompt);
                
                // This would call the actual AI API in production
                let result = StepResult::new(
                    true,
                    format!("{{\"content\": \"Generated content for: {}\", \"model\": \"{}\", \"temperature\": \"{}\"}}", 
                           prompt, model, temperature),
                    200,
                    "Content generated successfully".to_string()
                );
                self.step_results.insert(step_id, result);
            }
            "output" => {
                let data_ref = args.get(0).unwrap_or(&"data".to_string()).clone();
                let format = args.get(1).unwrap_or(&"text".to_string()).clone();
                let filename = args.get(2).unwrap_or(&"output".to_string()).clone();
                println!("    📤 Output: Export {} as {} to {}", data_ref, format, filename);
                
                let result = StepResult::new(
                    true,
                    format!("{{\"exported\": \"{}\", \"format\": \"{}\", \"file\": \"{}\"}}", 
                           data_ref, format, filename),
                    200,
                    "Output exported successfully".to_string()
                );
                self.step_results.insert(step_id, result);
            }
            "transform" => {
                let data_ref = args.get(0).unwrap_or(&"data".to_string()).clone();
                let transformation = args.get(1).unwrap_or(&"format".to_string()).clone();
                println!("    🔄 Transform: Apply {} to {}", transformation, data_ref);
                
                let result = StepResult::new(
                    true,
                    format!("{{\"transformed\": \"{}\", \"type\": \"{}\"}}", data_ref, transformation),
                    200,
                    "Data transformed successfully".to_string()
                );
                self.step_results.insert(step_id, result);
            }
            "validate" => {
                let data_ref = args.get(0).unwrap_or(&"data".to_string()).clone();
                let validation_type = args.get(1).unwrap_or(&"required".to_string()).clone();
                println!("    ✅ Validate: Check {} for {}", data_ref, validation_type);
                
                let result = StepResult::new(
                    true,
                    format!("{{\"validated\": \"{}\", \"type\": \"{}\", \"valid\": true}}", 
                           data_ref, validation_type),
                    200,
                    "Validation completed successfully".to_string()
                );
                self.step_results.insert(step_id, result);
            }
            _ => {
                println!("    ⚠️  Unknown command: {}", command.name);
                self.step_results.insert(step_id, StepResult::new(
                    false, "".to_string(), 400, format!("Unknown command: {}", command.name)
                ));
            }
        }
        
        Ok(())
    }
    
    fn execute_conditional(&mut self, conditional: &ConditionalStatement) -> Result<()> {
        let condition_result = self.evaluate_condition(&conditional.condition)?;
        
        if condition_result {
            println!("    ✅ Condition is true, executing if block");
            for step in &conditional.if_steps {
                self.execute_step(step)?;
            }
        } else {
            println!("    ❌ Condition is false");
            if let Some(else_steps) = &conditional.else_steps {
                println!("    🔄 Executing else block");
                for step in else_steps {
                    self.execute_step(step)?;
                }
            }
        }
        
        Ok(())
    }
    
    fn evaluate_condition(&self, condition: &Expression) -> Result<bool> {
        match condition {
            Expression::BinaryExpression { left, operator, right } => {
                let left_val = self.evaluate_expression(left)?;
                let right_val = self.evaluate_expression(right)?;
                
                match operator.as_str() {
                    "==" => Ok(left_val == right_val),
                    "!=" => Ok(left_val != right_val),
                    ">" => {
                        let left_num: f64 = left_val.parse().unwrap_or(0.0);
                        let right_num: f64 = right_val.parse().unwrap_or(0.0);
                        Ok(left_num > right_num)
                    }
                    "<" => {
                        let left_num: f64 = left_val.parse().unwrap_or(0.0);
                        let right_num: f64 = right_val.parse().unwrap_or(0.0);
                        Ok(left_num < right_num)
                    }
                    ">=" => {
                        let left_num: f64 = left_val.parse().unwrap_or(0.0);
                        let right_num: f64 = right_val.parse().unwrap_or(0.0);
                        Ok(left_num >= right_num)
                    }
                    "<=" => {
                        let left_num: f64 = left_val.parse().unwrap_or(0.0);
                        let right_num: f64 = right_val.parse().unwrap_or(0.0);
                        Ok(left_num <= right_num)
                    }
                    _ => Err(anyhow!("Unknown comparison operator: {}", operator)),
                }
            }
            _ => {
                let value = self.evaluate_expression(condition)?;
                Ok(!value.is_empty() && value != "0" && value != "false")
            }
        }
    }
    
    fn evaluate_expression(&self, expression: &Expression) -> Result<String> {
        match expression {
            Expression::StringLiteral(value) => Ok(value.clone()),
            Expression::NumberLiteral(value) => Ok(value.to_string()),
            Expression::Identifier(name) => {
                self.variables.get(name)
                    .cloned()
                    .ok_or_else(|| anyhow!("Undefined variable: {}", name))
            }
            Expression::BinaryExpression { left, operator, right } => {
                let left_val = self.evaluate_expression(left)?;
                let right_val = self.evaluate_expression(right)?;
                
                match operator.as_str() {
                    "+" => Ok(format!("{}{}", left_val, right_val)),
                    _ => Err(anyhow!("Unknown binary operator: {}", operator)),
                }
            }
            Expression::PropertyAccess { object, property } => {
                let object_val = self.evaluate_expression(object)?;
                // For now, just return the property name as a simple simulation
                Ok(format!("{}.{}", object_val, property))
            }
            Expression::StepReference { step_id, property } => {
                if let Some(result) = self.step_results.get(step_id) {
                    match property.as_deref() {
                        Some("status") => Ok(result.status.to_string()),
                        Some("data") => Ok(result.data.clone()),
                        Some("message") => Ok(result.message.clone()),
                        Some("success") => Ok(result.success.to_string()),
                        _ => Ok(result.data.clone()),
                    }
                } else {
                    Err(anyhow!("Step {} not found", step_id))
                }
            }
        }
    }
} 